### ISSUE TITLE
[M1] Adding new class hierarchy for new Disk Types

### ISSUE THREAD

**Issue by imaffe** [imaffe/external-workspace-manager-plugin#2]
> This is an unofficial issue format, just as part of  pre-view of official JIRA workflow.

For the old EWM, we only have `Disk` class with limited fields. To support EFS , we need to define other disk types. Hence we need to make `Disk` as abstract class/interfaces, and move current implementation of `Disk` to the derived class for extensibility (maybe other NFS provider will be added in the future).

We need to discuss how the refactor could be done, and list all the knowledges needed to do the refactor.

For new Disk types, like `AwsEfsDisk`, some new fields include :
1. AWS Credentials ID
2. VPC ID (as EC2 and EFS should be within the same VPC)
3. mount target ID

We can discuss on this, and start add codes including Web UI and backward-compatibility test.

**Comment by alexsomai** [comment]
I think this also depends on the strategy that we want to use.
Do we mount EFS on demand, as the workspace is being built? Or we mount it before actually building anything.

If we go for the second option, I think the current implementation may work. After mounting EFS, you can just populate the `Disk` fields as needed.

Otherwise, if we mount on demand, you'll also have to change the implementation of `ExwsAllocateExecution`, based on the current mounting config that you have.

**Comment by imaffe** [comment]
> I think this also depends on the strategy that we want to use.
> Do we mount EFS on demand, as the workspace is being built? Or we mount it before actually building anything.
> 
> If we go for the second option, I think the current implementation may work. After mounting EFS, you can just populate the `Disk` fields as needed.
> 
> Otherwise, if we mount on demand, you'll also have to change the implementation of `ExwsAllocateExecution`, based on the current mounting config that you have.

Agree, for the second option we do need to change the implementation of ExwsAllocateExecution and that is indeed a major part of the coding. Oleg says we would do it on-demand,  for if not on-demand there would be potentially too little code to work on. I'm OK with both design, and personally I'd think pre-mount instead of on-demand mounting is more close to actual use case, as these infrastructure should be managed by other CasC tools , and no one wants to mount/unmount NFS very frequently, as it will cost some speed overhead and unstablity.
